polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp

```diff
@@ -49,15 +50,94 @@ class SYCLRaiseHostConstructsPass
 // Pattern
 //===----------------------------------------------------------------------===//
 
+namespace {
+struct RaiseKernelName : public OpRewritePattern<LLVM::GlobalOp> {
+public:
+  using OpRewritePattern<LLVM::GlobalOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(LLVM::GlobalOp op,
+                                PatternRewriter &rewriter) const final {
+    // Get a reference to the kernel this global references
+    std::optional<SymbolRefAttr> ref = getKernelRef(op);
+    if (!ref)
+      return failure();
+
+    rewriter.replaceOpWithNewOp<sycl::SYCLHostKernelNameOp>(op, op.getSymName(),
+                                                            *ref);
+    return success();
+  }
+
+private:
+  /// If the input global contains a constant string representing the name of a
+  /// SYCL kernel, returns a reference to the `gpu.func` implementing this
+  /// kernel.
+  ///
+  /// The string will contain a trailing character we need to get rid of before
+  /// searching.
+  static std::optional<SymbolRefAttr> getKernelRef(LLVM::GlobalOp op) {
+    // Check the operation has a value
+    std::optional<Attribute> attr = op.getValue();
+    if (!attr)
+      return std::nullopt;
+
+    // Check it is a string
+    auto strAttr = dyn_cast<StringAttr>(*attr);
+    if (!strAttr)
+      return std::nullopt;
+
+    // Drop the trailing `0` character
+    StringRef name = strAttr.getValue().drop_back();
```

Comment:
Not sure about this but will this change the attribute on the global ? If so we need to put back the original unmodified string in the global so that if the global is not a kernel name (or the global is referenced in multiple instructions) the original input IR is unchanged.

Reply:
`getValue()` returns a `StringRef`, we are not modifying the original string, but generating a new `StringRef` by calling `drop_back`